src/main/java/org/wildfly/security/auth/realm/token/validator/JwtValidator.java

```diff
@@ -0,0 +1,304 @@
+        try {
+            Base64.Decoder urlDecoder = Base64.getUrlDecoder();
+            CodePointIterator decodedClaims = CodePointIterator.ofUtf8Bytes(urlDecoder.decode(encodedClaims));
+
+            return Json.createReader(decodedClaims.asUtf8().asInputStream()).readObject();
+        } catch (Exception cause) {
+            throw log.tokenRealmJwtParseFailed(cause);
+        }
+    }
+
+    private boolean verifySignature(String encodedHeader, String encodedClaims, String encodedSignature) throws RealmUnavailableException {
+        try {
+            Base64.Decoder urlDecoder = Base64.getUrlDecoder();
+            byte[] decodedSignature = urlDecoder.decode(encodedSignature);
+
+            boolean verify = ByteIterator.ofBytes(decodedSignature).verify(createSignature(encodedHeader, encodedClaims));
+
+            if (!verify) {
+                log.debug("Signature verification failed");
+            }
+
+            return verify;
+        } catch (Exception cause) {
+            throw log.tokenRealmJwtSignatureCheckFailed(cause);
+        }
+    }
+
+    private boolean hasValidAudience(JsonObject claims) throws RealmUnavailableException {
+        JsonValue audience = claims.get("aud");
+
+        if (audience == null) {
+            log.debug("Token does not contain an audience claim");
+            return false;
+        }
+
+        JsonArray audClaimArray;
+
+        if (JsonValue.ValueType.STRING.equals(audience.getValueType())) {
+            audClaimArray = Json.createArrayBuilder().add(audience).build();
+        } else {
+            audClaimArray = (JsonArray) audience;
+        }
+
+        boolean valid = audClaimArray.stream()
+                .map(jsonValue -> (JsonString) jsonValue)
+                .anyMatch(audience1 -> audiences.contains(audience1.getString()));
+
+        if (!valid) {
+            log.debugf("Audience check failed. Provided [%s] but was expected [%s].", audClaimArray.toArray(), this.audiences);
+        }
+
+        return valid;
+    }
+
+    private boolean hasValidIssuer(JsonObject claims) throws RealmUnavailableException {
+        String issuer = claims.getString("iss", null);
+
+        if (issuer == null) {
+            return false;
+        }
+
+        boolean valid = this.issuers.contains(issuer);
+
+        if (!valid) {
+            log.debugf("Issuer check failed. Provided [%s] but was expected [%s].", issuer, this.issuers);
+        }
+
+        return valid;
+    }
+
+    private Signature createSignature(String encodedHeader, String encodedClaims) throws NoSuchAlgorithmException, InvalidKeyException, SignatureException, RealmUnavailableException {
+        Signature signature = Signature.getInstance(resolveAlgorithm(encodedHeader));
+        Iterator<PemEntry<?>> pemEntryIterator = Pem.parsePemContent(CodePointIterator.ofUtf8Bytes(this.publicKey));
+        PublicKey publicKey = (PublicKey) pemEntryIterator.next().getEntry();
```

Comment:
Just noted something. I don't need to parse a PEM all the time. Will change that to use a previously built key instance instead.